common/src/main/java/org/fiware/tmforum/common/querying/QueryParser.java

```diff
@@ -0,0 +1,252 @@
+package org.fiware.tmforum.common.querying;
+
+import io.github.wistefan.mapping.JavaObjectMapper;
+import io.github.wistefan.mapping.NgsiLdAttribute;
+import io.github.wistefan.mapping.QueryAttributeType;
+import org.fiware.tmforum.common.exception.QueryException;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static org.fiware.tmforum.common.querying.Operator.GREATER_THAN;
+import static org.fiware.tmforum.common.querying.Operator.GREATER_THAN_EQUALS;
+import static org.fiware.tmforum.common.querying.Operator.LESS_THAN;
+import static org.fiware.tmforum.common.querying.Operator.LESS_THAN_EQUALS;
+import static org.fiware.tmforum.common.querying.Operator.REGEX;
+
+public class QueryParser {
+
+	// Keys for the "well-known" fields
+	public static final String OFFSET_KEY = "offset";
+	public static final String LIMIT_KEY = "limit";
+	public static final String FIELDS_KEY = "fields";
+	public static final String SORT_KEY = "sort";
+
+	public static final String NGSI_LD_OR = "|";
+	public static final String NGSI_LD_AND = ";";
+	// the "," in tm-forum values is an or
+	public static final String TMFORUM_OR_VALUE = ",";
+	// the ";" in tm-forum parameters is an or
+	public static final String TMFORUM_OR_KEY = ";";
+	public static final String TMFORUM_AND = "&";
+
+	public static boolean hasFilter(Map<String, List<String>> values) {
+		//remove the "non-filtering" keys
+		values.remove(OFFSET_KEY);
+		values.remove(LIMIT_KEY);
+		values.remove(FIELDS_KEY);
+		values.remove(SORT_KEY);
+		// if something is left, we have filter
+		return !values.isEmpty();
+	}
+
+	private static String removeWellKnownParameters(String queryString) {
+		List<String> parameters = Arrays.asList(queryString.split(TMFORUM_AND));
+		List<String> wellKnownParams = parameters
+				.stream()
+				.filter(p -> p.startsWith(LIMIT_KEY)
+						|| p.startsWith(FIELDS_KEY)
+						|| p.startsWith(OFFSET_KEY)
+						|| p.startsWith(SORT_KEY)
+				)
+				.toList();
+		// not part of the query
+		parameters.removeAll(wellKnownParams);
+		return String.join(TMFORUM_AND, parameters);
+	}
+
+	public static String toNgsiLdQuery(Class<?> queryClass, String queryString) {
+
+		queryString = removeWellKnownParameters(queryString);
+
+		List<String> parameters;
+		LogicalOperator logicalOperator = LogicalOperator.AND;
+		// tmforum does not define queries combining AND and OR, therefor we
```

Comment:
Shall we validate that the user does not violate the spec by using both?

Reply:
Yes, makes sense to explicitly check it. I'll add